Readme.md
<div align="center">

![Logo](docs/_images/logos/logo_100x100.png)

# [Vayu](https://en.wikipedia.org/wiki/Vayu)

** [Vayu](https://en.wikipedia.org/wiki/Vayu) is a framework built on top of pytorch for training binary classification models. **

[![pipeline status](https://gitlab.qpidhealth.net/qaas/deep-learning/badges/master/pipeline.svg)](https://gitlab.qpidhealth.net/qaas/deep-learning/-/commits/master)
[![coverage report](https://gitlab.qpidhealth.net/qaas/deep-learning/badges/master/coverage.svg)](https://gitlab.qpidhealth.net/qaas/deep-learning/-/commits/master)

</div>




 
 Support for named entity
recognition (NER), language and knowledge graph models is planned for future releases which are now standalone scripts.


### Docs
- [latest](http://10.205.63.29/docs/vayu/latest) 
- [stable](http://10.205.63.29/docs/vayu/stable)
- [0.1.1](http://10.205.63.29/docs/vayu/0.1.1)
- [0.1.0](http://10.205.63.29/docs/vayu/0.1.0)


### Requirements

This project requires Python 3.6+

Create a conda environment (preferred)
```bash
conda env create --file conda_requirements.yml
conda activate vayu
```


### Data
* JsonL : One line per record in json format exported using following scala script using spark
* Having metadata columns like `cptCode` and `contactMethod` is helpful if present to export `model_info_advanced.json`
* The tokens column is `List[List[str]]` / Doc[sentences] where doc is List[sentence], sentence is `List[str]`
* Datasets should already be split into train, validation and test in (BIO CoNLL format for NER) for the scripts to
 ingest. `vayu/datasets/train_valid_split.py` can be used to split datasets.
* Having a static validation and test set instead of random is important for measure performance of new models
 compared to earlier ones. 

### Train

`python -m main do_train=true`
1. Training flow by default trains -> tunes -> evaluates
2. By default, training gets configured using yaml situated in `conf/config.yaml`. 
3. To override any aspect of training, either pass via cmdline e.g. `python -m main training.optimizer_learning_rate=0.1 model=classification/cnn_lstm model.params.embedding_size=10`
4. Configs are composed from multiple hierarchical small config files along with their defaults. This allows highly
 flexible configuration e.g. having same training, data params but different params for different models.
5. After training, system outputs the path of trained model which consists of: 


    a. complete config.yaml
    b. lightning_logs/version_0 will have by default top 3 models found during training based on validation loss.
    c. model_info_advanced.json, thresholds.json, test_results.yaml, tune_results.yaml as well as train.log which
     comprises of stdout of training.

### Hyperparameter sweeps
```
python -m main do_train=true --multirun \ 
model.params.embedding_size=10,50,100 \ 
hydra.sweep.dir=multiruns/embedding_effect \ 
hydra.sweep.subdir=\${model.embedding_size}
```
Above will create separate directories for each embedding_size, this saves developer effort in running multiple configs.
     
### Resume training from checkpoint
```
python -m main do_train=true \
model=classification/boe \
pl_trainer.checkpoint_path=path/to/checkpoint
```

### Evaluate a pretrained model
You can evaluate on a directory of json files or single file
* This assumes your config is present in `outputs/2020-05-27/16-28-42/.hydra/config.yaml`
* If you want to evaluate model on a separate threshold, simply update `threshold` param in `thresholds.json`
* If you want to tune again on different dataset:
    1. backup and delete `thresholds.json` in pretrained model dir.
    2. If `thresholds.json` is not deleted, then it will be used for prediction.
    3. New threshold will be exported to a new directory to avoid overwriting.
```bash
python -m main do_eval=true do_train=false \ 
pretrained_model_path=outputs/2020-05-27/16-28-42 \
data.test_path=path \  # path can be dir of files or single json file
data.val_path=path  # if you want to tune on a new dataset
```



### Tensorboard
Here you can compare experiments 
```bash
tensorboard --logdir="outputs"
```

### Todo
* [x] Does not support batch size > 1 when training throws OOM for a bigger batch size, if you lower max_chunks and
 chunk_size which will support larger batch size
    * [x] Custom collate functions for Dataloader
* [ ] Categorical data featurization (pytorch-tabnet)
* [x] Support for streaming json datasets loading w/o blowing up memory
* [x] Support for loading multiple json datasets
* [x] Training takes evaluation set as param from beginning, which is a limitation for now since we would want to evaluate on any dataset post training. Workaround for that is load model manually in an interpreter and run custom evaluate dataset through it
* [ ] Imbalanced batch sampler
 
### Milestones
- [x] CNN + classification head
- [x] CNN + LSTM + classification head
- [x] HAN + classification head
- [ ] CNN + HAN + classification head
- [x] Support gensim pretrained vectors (w2v, d2v, fastText and glove)
- [x] Transformers + Classification Head
- [ ] CDR custom tokenizer
- [ ] Hybrid CDR tokenizer that leverages a whitelist along with BPE/wordpiece tokenization

### Core libraries
- [PyTorch](https://pytorch.org/docs/stable/index.html)
- [PyTorch-Lightning](https://pytorch-lightning.readthedocs.io/en/latest/)
- [Hydra](https://hydra.cc)
- [Transformers](https://huggingface.co/transformers/index.html)


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[v0.2.0] - 11/11/2020
---
### Added
* Scripts to register results of any model as a run in any AML experiment [source MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/21)
* Ability to load embeddings using azure ml dataset [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Calibration on training set by default [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Log metrics to Azure ML [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Output calibration set predictions on optimal threshold [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* AML configuration and scripts to train models on azure
* Experiment tags [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* RunID in slack message upon experiment completion [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)

### Changed
* Copy embedding to model dir on serialization [!1](https://inqbator-gitlab.innovate.lan/singhay/vayu/-/merge_requests/1)
* Calibration tqdm update progress bar every 1/10th time [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Avoid uploading accuracy tables because they don't load in UI [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Moved aml.docker to respective cpu, gpu configs [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Renamed `val_path` to `valid_path` [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)

### Removed
* Gensim dependency, instead use numpy matrix to load pretrained embeddings [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)
* Unwanted nested checkpoint saving dirs `lightning_logs/version_0` [AML177 MR](https://gitlab.qpidhealth.net/qaas/vayu/-/merge_requests/20)

[v0.1.1] - 07/15/2020
---
### Added
- Label smoothing
- Label squeezing
- Add new model that uses 1d convolution instead of 2d increasing performance by 46% and reduced training time from 3h to 2h for 360k records.
- Auto build sphinx docs and push to web server using CI
- Add gitlab badges for test coverage

### Changed
- Make tuning on training set default instead of tuning set
- Moved tests directory outside of source directory
- Moved one-off scripts into their own "scripts" module from "vayu"

### Removed
- Configurable truth configs because they are detrimental to performance
- Task yaml because it was not needed
- requirements.txt



main.py
import logging
import os
import shutil

import hydra
import pytorch_lightning as pl
import torch
from azureml.core import Run
from hydra import utils
from omegaconf import OmegaConf, DictConfig
from pytorch_lightning import seed_everything
from pytorch_lightning.callbacks import EarlyStopping, ModelCheckpoint
from torch.backends import cudnn
from torch.utils.data import SequentialSampler, DataLoader

from vayu.callbacks import PersistModelProperties, AMLogger
from vayu.constants import Split, THRESHOLDS_FILE
from vayu.models.cdr_lightning_mixin import CDRLightningMixin
from vayu.models.classification.optimal_threshold_finder import OptimalThresholdFinder
from vayu.score_calculations import ScoreCalculations
from vayu.utils import get_top_checkpoint_path

use_cuda = torch.cuda.is_available()
device = torch.device("cuda:0" if use_cuda else "cpu")
cudnn.benchmark = True

logger = logging.getLogger(__name__)


def train(model: CDRLightningMixin, cfg: dict, run: Run) -> None:
    """
    Main training routine specific for this project
    :param model: child of CDRLightningMixin
    :param cfg: dictionary of hyperparameters
    :param run: azureml run object
    """
    # ------------------------
    # INIT CALLBACKS
    # ------------------------
    early_stop_callback = EarlyStopping(**cfg['callbacks']['early_stop'])

    # ------------------------
    # INIT TRAINER
    # ------------------------
    trainer = pl.Trainer(**cfg['pl_trainer'],
                         early_stop_callback=early_stop_callback if cfg['is_early_stop'] else None,
                         resume_from_checkpoint=cfg['pl_trainer']['checkpoint_path'] or None,
                         callbacks=[PersistModelProperties(cfg['output_dir'])],
                         logger=AMLogger(cfg['output_dir'], run),
                         reload_dataloaders_every_epoch=cfg['data']['active_learning']['is_enable']
                         )

    # --------------------------------
    # INIT MODEL CHECKPOINT CALLBACK
    # -------------------------------
    ckpt_path = os.path.join(
        trainer.default_save_path,
        "checkpoints-{epoch:03d}-{val_loss:.3f}-{loss:.3f}",
    )
    # initialize Model Checkpoint Saver
    checkpoint_callback = ModelCheckpoint(
        filepath=ckpt_path,
        **cfg['callbacks']['checkpoint']
    )
    trainer.checkpoint_callback = checkpoint_callback

    # ------------------------
    # START TRAINING
    # ------------------------
    trainer.fit(model)

    # ------------------------
    # START TUNING
    # ------------------------
    dataloader = model.train_dataloader() if cfg['is_calibrate_on_train'] else model.val_dataloader()
    # dataloader = read_data_from_local_file(model, cfg['data']['train_path'], Split.TRAIN)
    # Load best model from checkpoint
    loss, model_path = get_top_checkpoint_path(cfg['output_dir'])
    logger.info("Loading model from %s", model_path)
    model = model.load_from_checkpoint(model_path,
                                       hparams_file=os.path.join(trainer.weights_save_path, "hparams.yaml"),
                                       map_location=torch.device('cpu') if cfg['n_gpus'] == 0 else device,
                                       **cfg['training']['hparams'])

    calibration_set_type = Split.TRAIN if cfg['is_calibrate_on_train'] else Split.VALID
    ids, predictions, targets, features = predict(model, dataloader, cfg['n_gpus'],
                                                  calibration_set_type.value + "_calibration_set")

    otf = OptimalThresholdFinder(**cfg['callbacks']['threshold'])
    otf.calculate_thresholds(predictions, targets)
    otf.export_thresholds(output_dir=cfg['output_dir'])

    scores = ScoreCalculations(predictions, targets, otf, calibration_set_type)
    scores.export_metrics_to_yaml(cfg['output_dir'])
    scores.export_predicted_probas_to_csv(cfg['output_dir'], ids=ids, features=features,
                                          is_export_features=cfg['is_export_features'])

    # ------------------------
    # START EVALUATING
    # ------------------------
    evaluate(model, cfg['data']['test_path'], cfg['output_dir'], otf, cfg['n_gpus'], run, cfg['is_export_features'])


def evaluate(model, test_data_path: str, test_output_dir: str,
             otf: OptimalThresholdFinder, n_gpus: int, run: Run, is_export_features: bool) -> None:
    logger.info("Starting evaluation")

    ids, predictions, targets, features = [], [], [], []

    # If single file is passed, the individual eval is already complete
    if os.path.isdir(test_data_path):
        # The following code outputs a yaml and predictions csv in a directory
        if len(os.listdir(test_data_path)) > 0:
            individual_evals_out_dir_path = os.path.join(test_output_dir, "individual_evals")
            os.mkdir(individual_evals_out_dir_path)
            for eval_jsonl_file_name in os.listdir(test_data_path):
                eval_jsonl_path = os.path.join(test_data_path, eval_jsonl_file_name)
                logger.info("Starting evaluation for %s", eval_jsonl_path)
                file_name, extension = os.path.splitext(eval_jsonl_file_name)

                dataloader = read_data_from_local_file(model, eval_jsonl_path, Split.TEST)
                ids_indiv, predictions_indiv, targets_indiv, features_indiv = predict(model,
                                                                                      dataloader,
                                                                                      n_gpus, file_name)
                scores = ScoreCalculations(predictions_indiv, targets_indiv, otf, Split.TEST)

                scores.export_metrics_to_yaml(individual_evals_out_dir_path, file_name + "_")
                scores.export_predicted_probas_to_csv(individual_evals_out_dir_path, ids_indiv,
                                                      features_indiv, file_name + "_", is_export_features)

                ids.extend(ids_indiv)
                predictions.extend(predictions_indiv)
                targets.extend(targets_indiv)
                features.extend(features_indiv)
        else:
            raise ValueError(f"{test_data_path} is empty, please pass a directory"
                             "that contains jsonl files for evaluating individually.")
    else:
        test_dataloader = read_data_from_local_file(model, test_data_path, Split.TEST)
        ids, predictions, targets, features = predict(model, test_dataloader, n_gpus,
                                                      os.path.basename(test_data_path))

    scores = ScoreCalculations(predictions, targets, otf, Split.TEST)
    scores.export_metrics_to_yaml(test_output_dir)
    scores.export_predicted_probas_to_csv(test_output_dir, ids=ids, features=features,
                                          is_export_features=is_export_features)
    scores.log_metrics_to_aml(run)


def read_data_from_local_file(model, data: str, data_split_type: Split):
    dataset = model.dataset_cls(data=data, data_split_type=data_split_type,
                                cat_encoding=model.cat_encoding)
    dataloader = model.dataloader_cls(dataset, sampler=SequentialSampler(dataset))

    return dataloader


def predict(model, dataloader: DataLoader, n_gpus: int, tqdm_desc: str):
    if n_gpus != 0:
        model.cuda()
    model.freeze()
    # multi-gpu evaluate
    if n_gpus > 1 and not isinstance(model, torch.nn.DataParallel):
        model = torch.nn.DataParallel(model)
    model.eval()

    ids, predictions, targets, features = [], [], [], []
    # for batch in tqdm(dataloader, desc=tqdm_desc, miniters=round(0.1 * len(dataloader))):
    logger.info("Predicting on %s", tqdm_desc)
    for batch in dataloader:
        batch = [batch[0]] + [data.to(device) for data in batch[1:]]
        idxs, logits, labels, encodings = model(**dict(batch=batch, is_return_features=True))
        ids.extend(idxs)
        targets.extend(labels.detach().cpu().tolist())
        predictions.extend(torch.sigmoid(CDRLightningMixin.squeeze(logits)).detach().cpu().tolist())
        features.extend([','.join(map(str, encoding)) for encoding in encodings.detach().cpu().tolist()])

    return ids, predictions, targets, features


def get_dataset(input_datasets: dict, dataset_type: str) -> str:
    """Fetches the local path of dataset from all the datasets in AML run

    Args:
        input_datasets (dict): A dictionary of input datasets from Azure run
        dataset_type (str): The type of dataset (either of train, tune or test)

    Returns:
        str: The path of dataset in local filesystem
    """
    for dataset_name, dataset_path in input_datasets.items():
        if dataset_type in dataset_name:
            return dataset_path
    raise ValueError(f"{dataset_type} not found in datasets")


@hydra.main(config_path="conf/config.yaml", strict=False)
def main(cfg: DictConfig) -> None:
    # So that model, lightning and hydra logs are written to the same place
    cfg.output_dir = os.getcwd()
    cfg.n_gpus = torch.cuda.device_count()
    seed_everything(cfg.seed)

    # Setup logging
    logging.basicConfig(
        format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
        datefmt="%m/%d/%Y %H:%M:%S",
        level=logging.INFO
    )
    logger.warning(
        "device: %s, n_gpu: %s, 16-bits training: %s",
        device,
        cfg.n_gpus,
        cfg.use_fp16
    )

    # ------------------------
    # DATA LOADING
    # if not os.path.exists(cfg.data.val_path):  # TODO: data validation checks
    #    raise FileNotFoundError(f"No validation set found at {cfg.data.val_path}")
    # ------------------------
    # Get the experiment run context
    run = Run.get_context(allow_offline=True)
    if cfg.is_azure_train:
        cfg.data.train_path = get_dataset(run.input_datasets, Split.TRAIN.value)
        cfg.data.valid_path = get_dataset(run.input_datasets, Split.VALID.value)
        cfg.data.test_path = get_dataset(run.input_datasets, Split.TEST.value)
        cfg.data.tokenizer_path = utils.to_absolute_path(cfg.data.tokenizer_path)
        if 'pretrained_vector_path' in cfg.model.params and cfg.model.params.pretrained_vector_path != '':
                cfg.model.params.pretrained_vector_path = get_dataset(run.input_datasets, Split.EMBEDDING.value)

        if cfg.data.cat_encoding:
            for key in cfg.data.cat_encoding:
                path = cfg.data.cat_encoding[key]
                cfg.data.cat_encoding[key] = get_dataset(run.input_datasets, path)

    if 'pretrained_vector_path' in cfg.model.params and cfg.model.params.pretrained_vector_path != '':
        shutil.copytree(cfg.model.params.pretrained_vector_path,
                        os.path.join(cfg.output_dir, Split.EMBEDDING.value))
        cfg.model.params.pretrained_vector_path = os.path.join(cfg.output_dir, Split.EMBEDDING.value)

    if cfg.pretty_print:
        print(cfg.pretty())
    # ------------------------
    # INIT MODEL
    # ------------------------
    cfg.training.hparams.merge_with(cfg.model.params, cfg.data)

    if cfg.do_train:
        model = hydra.utils.instantiate(cfg.model,
                                        **cfg.training,  # redundant but required by lightning module
                                        **cfg.training.hparams)
        train(model, OmegaConf.to_container(cfg, resolve=True), run)
    elif cfg.do_eval_pretrained_model:
        # ------------------------
        # LOAD MODEL
        # Load model config
        # Load best model from checkpoints with lowest validation loss
        # ------------------------
        cfg.pretrained_model_path = utils.to_absolute_path(cfg.pretrained_model_path)
        # TODO: add ability to evaluate multiple checkpoints
        loss, model_path = get_top_checkpoint_path(cfg.pretrained_model_path)
        logger.info("Loading model from %s", model_path)

        pretrained_config_path = os.path.join(cfg.pretrained_model_path, '.hydra/config.yaml')
        if os.path.exists(pretrained_config_path):
            # TODO: this part needs an integration test so bad
            pretrained_config = OmegaConf.load(pretrained_config_path)
            pretrained_config.training.hparams.merge_with(pretrained_config.model.params,
                                                          pretrained_config.data)
            model = hydra.utils.instantiate(pretrained_config.model,
                                            **pretrained_config.training,  # redundant but required by lightning module
                                            **pretrained_config.training.hparams)
            model = model.load_from_checkpoint(model_path,
                                               map_location=torch.device('cpu') if
                                               cfg.n_gpus == 0 else None,
                                               **pretrained_config.training.hparams)
            # This is so we can support command line argument overrides
            cfg = OmegaConf.merge(pretrained_config, cfg)
        else:
            raise FileNotFoundError("No config file found at `.hydra/config.yaml`")

        # ----------------------------------------
        # START TUNING
        # Load threshold file if present else tune
        # ----------------------------------------
        logger.info("Starting tuning")
        if os.path.exists(os.path.join(cfg.pretrained_model_path, THRESHOLDS_FILE)):
            otf = OptimalThresholdFinder.from_precomputed_thresholds(cfg.pretrained_model_path)
        else:
            if not os.path.exists(cfg.data.val_path):
                raise FileNotFoundError(f"No validation set found at {cfg.data.val_path}")

            predictions, targets = predict(model, cfg.data.val_path, cfg.n_gpus, Split.VALID.value)
            otf = OptimalThresholdFinder(**cfg.callbacks.threshold)
            otf.calculate_thresholds(predictions, targets)
            otf.export_thresholds(output_dir=cfg.output_dir)

        # ------------------------------------------------
        # START EVALUATING
        # Write predictions to a csv and metrics to a yaml
        # ------------------------------------------------
        evaluate(model, cfg.data.test_path, cfg.output_dir, otf, cfg.n_gpus, run, is_export_features=True)


def set_environment_variables_for_nccl_backend(single_node=False, master_port=6105):
    if not single_node:
        master_node_params = os.environ["AZ_BATCH_MASTER_NODE"].split(":")
        os.environ["MASTER_ADDR"] = master_node_params[0]

        # Do not overwrite master port with that defined in AZ_BATCH_MASTER_NODE
        if "MASTER_PORT" not in os.environ:
            os.environ["MASTER_PORT"] = str(master_port)
    else:
        os.environ["MASTER_ADDR"] = os.environ["AZ_BATCHAI_MPI_MASTER_NODE"]
        os.environ["MASTER_PORT"] = "54965"

    os.environ["NCCL_SOCKET_IFNAME"] = "^docker0,lo"
    os.environ["NODE_RANK"] = os.environ[
        "OMPI_COMM_WORLD_RANK"
    ]  # node rank is the world_rank from mpi run


if __name__ == "__main__":
    # set_environment_variables_for_nccl_backend(single_node=False)
    main()
